[v1.9] Fix image digest preparation for release commits

joestringer opened the pull request:
- install: add digests into helm charts
- install/kubernetes: do not use digests

Partially backports PRs #15185 and #15318.

aanm committed:
install/kubernetes: do not use digests

[ upstream commit 5ca7c948717539ff168b7f340fa01eb5408d4df7 ]

[ Backporter's notes: Changed 'use_digest' to false by default in the
                      update-versions make target, to avoid forcing
		      existing users into using digests during upgrade;
                      Concerned users can opt in manually by specifying
                      the use digest options in the helm args/config. ]

As digests are not available when prepare a release, digests should not
be set. Once the release is performed and Makefile.digests contains
those digests, we will automatically use the digests available.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -41,7 +41,7 @@ EXPERIMENTAL_OPTIONS := \
     --set operator.replicas=1
 
-USE_DIGESTS ?= true
+USE_DIGESTS ?= $(shell if grep -q '""' Makefile.digests; then echo "false"; else echo "true"; fi)
```
I'm actually not 100% sure if this is what we intend for the v1.8/v1.9 branches, maybe this is just supposed to be `USE_DIGESTS ?= false`. Then anyone on those releases can opt in for newer helm releases.

aanm replied:
you meant "newer image digests"? If not, then I didn't understand what you were trying to explain here.

joestringer replied:
> I think that we have discussed that this should be false by default for stable branches to avoid breaking any users deployment where people have their own images.

This is what I mean. + opt in.

aanm commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -1393,7 +1393,7 @@ clustermesh:
       # clustermesh-apiserver-digest
       digest: "sha256:e4a6d47d0d8e4ef52c44e99a906076e29e8513a35a972e60100df2168b381318"
-      useDigest: false
+      useDigest: true
```
I think that we have discussed that this should be false by default for stable branches to avoid breaking any users deployment where people have their own images.

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -89,16 +89,19 @@ update-versions:
 		branch="$$cilium_version";											\
 		hubble_version=$(HUBBLE_UI_VERSION);										\
 		pull_policy="IfNotPresent";											\
+		use_digest="false";											\
```
This was 'true' in the original commit, but I've changed it to 'false' in this PR so that the helm charts disable digests by default.

As a result, generated helm charts will have them disabled; however, the experimental / quick installs will enable them by default using the result of `USE_DIGEST` at the top of this commit. I believe this is consistent with how the branch is currently operating, but with one extra improvement: When preparing the release commit for the next release (when there are no valid digests), rather than setting the image tags in such generated YAMLs (used in smoke tests) falsely to `cilium/cilium:latest@`, it will instead generate `cilium/cilium:latest`. After we fill in the digests post-release, the quick install will point to actual image shas for the new release.

aanm replied:
shouldn't point to `cilium/cilium:v1.9` instead?

joestringer replied:
`cilium/cilium:v1.9.6` or similar, yeah. `latest` was a typo on my side, my intent was to emphasize the extra `@` which breaks the test.

joestringer commented on install/kubernetes/Makefile:
```diff
@@ -116,12 +119,13 @@ update-versions:
-		sed -i '/# operator-generic-digest.*/{!b;n;s/genericDigest.*/genericDigest: "'$(OPERATOR_GENERIC_DIGEST)'"/}' $(CILIUM_VALUES); \
+		sed -i '/# operator-aws-digest.*/{!b;n;s/aws-digest.*/aws-digest: "'$(OPERATOR_AWS_DIGEST)'"/}' $(CILIUM_VALUES); 		\
+		sed -i '/# operator-azure-digest.*/{!b;n;s/azure-digest.*/azure-digest: "'$(OPERATOR_AZURE_DIGEST)'"/}' $(CILIUM_VALUES); 	\
+		sed -i '/# operator-generic-digest.*/{!b;n;s/generic-digest.*/generic-digest: "'$(OPERATOR_GENERIC_DIGEST)'"/}' $(CILIUM_VALUES); \
```
:thinking: this actually looks wrong to me, I think we may have broken this on master.

joestringer replied:
OK yeah we already fixed it on master (eg commit 8d5c4cb24edc5dec090dd367a7a2d7d28de53dcb). I just need to fix this version to revert this hunk.

joestringer committed:
install: add digests into helm charts

[ upstream commit 4ebe592602f730c03f47dd345b9ba94896d0fcf3 ]

[ Backporter's notes: Discard all changes outside the Makefile; don't
                      apply the changes that were subsequently fixed in
                      commit 8d5c4cb24edc5. ]

This commit adds a simple way to add the image digests into the official
helm charts.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Joe Stringer <joe@cilium.io>
